### ISSUE TITLE
no-undefined-types: False positives with class templates

### ISSUE THREAD

**Issue by l1bbcsg** [gajus/eslint-plugin-jsdoc#363]
Test case 
```js
    {
      code: `
      /**
       * @template TEMPLATE_TYPE
       */
      class Foo {
        /**
         * @return {TEMPLATE_TYPE}
         */
        passes () {}
      
        /**
         * @return {TEMPLATE_TYPE}
         */
        fails () {}
      }
      `,
    },
```

The first method (`passes()`) does not raise any erroros while the second one (`fails()`) does.
The reason for this is that `getClassJsdoc()` returns null for the second method.

This was introduced in a1c9824be2bbbf1766e04145a4e4c7a52bd9935d. Namely, `getAncestor` implementation walks through tokens rather than AST nodes and happens to work for the first node because the third token is `class` keyword while for the second method it's a `{` punctuator.

Another way to break this would be declaring the class as `class Foo extends Bar`, this way the third token from the node would be `Foo` identifier. There should be plenty of other ways as well.

This potentially affects all calls to `getClassJsdoc` and `getClassNode` so it might have impact on some other rules as well.

**Comment by golopot** [comment]
Why do you want to put template tag in a class node? Template tag should be put in the function node.

**Comment by l1bbcsg** [comment]
No, this is the correct GCC [Generic Types](https://github.com/google/closure-compiler/wiki/Generic-Types) syntax.

**Comment by golopot** [comment]
I got you. This should be recognized indeed.

**Comment by brettz9** [comment]
The difficulty here is that we need the [`Program`](https://github.com/gajus/eslint-plugin-jsdoc/blob/master/src/iterateJsdoc.js#L415) hook because the `Program:exit` hook wasn't reporting for the comment block above the likes of a global `var quux = {};`, and with the `Program` hook, we apparently can't get access to the AST ancestors here.

**Comment by l1bbcsg** [comment]
I looked into it more and I believe neither of `Program` or `Program:exit` hooks fix this. Namely, to get class node we need to access AST ancestors of AST node jsdoc comment is associated with. However, when rules are iterating over `Program` the only node they're getting is the Program itself which doesn't even have any ancestors in the first place, so ancestors lookup was fundamentally broken.

The only way I figured how to fix this is to actually iterate all nodes rather than comments and attempted this in #456.
